File: gcpc2025/bustlingbusride/submissions/accepted/yidi.py
n, b, delay, w = map(int, input().split())
a = [*map(int, input().split())]
for i in range(1, b):
    a[i] += a[i-1]
c = [*map(int, input().split())]
for i in range(n):
    c[i] -= 1

def upd(seg, i, x):
    i += b
    seg[i] += x
    while i > 1:
        i //= 2
        seg[i] = seg[2*i] + seg[2*i+1]

def query(seg, l, r):
    ans = 0
    l += b
    r += b
    while l < r:
        if l&1:
            ans += seg[l]
            l += 1
        if r&1:
            r -= 1
            ans += seg[r]
        
        l //= 2
        r //= 2

    return ans

seg1 = [0] * (2*b)
seg2 = [0] * (2*b)
used = [0] * b

l = 0
r = 2 * 10 ** 18
while l+1 < r:
    mid = (l+r) // 2
    ptr = 0
    bus = 0
    while ptr < n:
        t = bus * delay
        bus += 1
        if t + 2*w + c[ptr] > mid:
            break

        mx = 0
        rem = ptr
        while ptr < n:
            mx = max(mx, a[c[ptr]])
            t += query(seg1, 0, c[ptr]) * w * 2
            if not used[c[ptr]]:
                t += query(seg2, c[ptr]+1, b) * w * 2
                upd(seg1, c[ptr], 1)
                used[c[ptr]] = 1
            upd(seg2, c[ptr], 1)
            t += 2*w
            if t + mx > mid:
                break

            ptr += 1

        while rem < n and rem <= ptr:
            if used[c[rem]]:
                upd(seg1, c[rem], -1)
            used[c[rem]] = 0
            upd(seg2, c[rem], -1)
            rem += 1

    if ptr < n:
        l = mid
    else:
        r = mid

print(r)

File: gcpc2025/bustlingbusride/submissions/accepted/yidi.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define all(x) ::begin(x), ::end(x)
void _d(auto... x) { ((cerr << ' ' << x), ...) << endl; }
#define debug(x...) cerr << "["#x"]:", _d(x)


#include <ext/pb_ds/assoc_container.hpp>
using namespace __gnu_pbds;
template<typename T>
using Tree = tree<T, null_type, less<T>, rb_tree_tag,
                  tree_order_statistics_node_update>;

int main() {
    cin.tie(0)->sync_with_stdio(0);

    int n, b, delay, w;
    cin >> n >> b >> delay >> w;
    vector<int> a(b);
    for (int& e : a) cin >> e;
    for (size_t i = 1; i < b; i++) a[i] += a[i-1];
    vector<int> c(n);
    for (int& e : c) cin >> e, e = a[e-1];

    constexpr ll INF = 2e18;
    ll l = 0, r = INF;
    while (l+1 < r) {
        ll mid = (l+r) / 2;
        int ptr = 0;
        for (int bus=0; ptr<n; bus++) {
            ll t = (ll)bus*delay;
            if (t + 2*w + c[ptr] > mid) break;

            Tree<int> tree;
            Tree<pair<int, int>> tree2;
            int mx = 0;
            while (ptr < n) {
                mx = max(mx, c[ptr]);
                t += (tree.order_of_key(c[ptr])) * w * 2;
                if (tree.find(c[ptr]) == tree.end()) {
                    t += ((ll)tree2.size() - tree2.order_of_key({c[ptr], ptr})) * w * 2;
                    tree.insert(c[ptr]);
                }
                tree2.insert({c[ptr], ptr});
                t += 2*w;
                if (t + mx > mid) break;
                ptr++;
            }
        }
        if (ptr < n) l = mid;
        else r = mid;
    }
    cout << r << '\n';
}
